mmv1/third_party/terraform/services/container/go/resource_container_cluster.go.tmpl

```diff
@@ -4252,6 +4253,27 @@ func resourceContainerClusterUpdate(d *schema.ResourceData, meta interface{}) er
 		}
 	}
 
+	if d.HasChange("node_pool_defaults") && d.HasChange("node_pool_defaults.0.node_config_defaults.0.insecure_kubelet_readonly_port_enabled") {
```

Comment:
Based on my reading of the logs for the remaining failing test, I don't think think this is making an update when the value in the config changes from `true` to `false`.

Reply:
Yes, agree, and that's what shows up in my functional testing as well (and, as mentioned below, I'm seeing a similar permadiff type problem with `node_config` as well, when changing the existing value, even though that test is passing):
```
      - .node_config[0].kubelet_config[0].insecure_kubelet_readonly_port_enabled: was cty.False, but now cty.True
```

Or am I missing a critical component or screwing up the data structure somewhere?

FWIW, this is how everything looks at the API level. I implemented it similarly to the logging variant one where it _doesn't_ totally mirror the structure (for `node_pool_defaults`, i.e., there's no `kubelet_config` block there).

```yaml
nodeConfig:
  # [...]
  kubeletConfig:
    cpuCfsQuota: false
    cpuManagerPolicy: static
    insecureKubeletReadonlyPortEnabled: true
# [...]
nodePoolDefaults:
  nodeConfigDefaults:
    loggingConfig:
      variantConfig:
        variant: DEFAULT
    nodeKubeletConfig:
      insecureKubeletReadonlyPortEnabled: true
```

One thing I did find was a duplicate bit for `insecure_kubelet_readonly_port_enabled` in `expandNodeConfig()`, where that should actually come from the nested `expandKubeletConfig()` within it - will push up that update and the other update.

Reply:
Btw, this is the simple test I'm using locally (switching between true and false, obviously) if it's at all helpful

```hcl
resource "google_container_cluster" "test_insecure_kubelet_readonly_port" {
  name               = "test-insecure-kubelet-readonly-port"
  location           = "us-central1-f"
  initial_node_count = 1
  
  node_config {
    kubelet_config {
      cpu_manager_policy                     = "static"
      insecure_kubelet_readonly_port_enabled = false
    }
  }

  node_pool_defaults {
    node_config_defaults {
      insecure_kubelet_readonly_port_enabled = false
    }
  }

  deletion_protection = false
  network             = "default"
  subnetwork          = "default"
}
```

Reply:
Confirmed that making a cluster with both set to false, and then toggling both to true, it _does_ work for `node_pool_defaults` (but not `node_config`), but then doesn't work going the other direction. Also confirmed that the value is unset in the API when creating a new cluster with the value set to `false`. So maybe we do need some kind of custom flatten function to handle the value when the API isn't returning those fields since they're nested?

```
2024-08-12T11:25:21.749-0700 [DEBUG] provider.terraform-provider-google: {
2024-08-12T11:25:21.749-0700 [DEBUG] provider.terraform-provider-google:  "update": {
2024-08-12T11:25:21.749-0700 [DEBUG] provider.terraform-provider-google:   "desiredNodeKubeletConfig": {
2024-08-12T11:25:21.749-0700 [DEBUG] provider.terraform-provider-google:    "insecureKubeletReadonlyPortEnabled": true
2024-08-12T11:25:21.749-0700 [DEBUG] provider.terraform-provider-google:   }
2024-08-12T11:25:21.749-0700 [DEBUG] provider.terraform-provider-google:  }
2024-08-12T11:25:21.749-0700 [DEBUG] provider.terraform-provider-google: }
```

Reply:
ps - even though we know `NodeConfigDefaults` is not `nil` already, I don't think we know whether `NodeConfigDefaults.NodeKubeletConfig` is or not?